Source/Core/Common/WorkQueueThread.h

```diff
@@ -27,21 +27,34 @@ class WorkQueueThread
   {
     Shutdown();
     m_shutdown.Clear();
+    m_cancelled.Clear();
     m_function = std::move(function);
     m_thread = std::thread(&WorkQueueThread::ThreadLoop, this);
   }
 
   template <typename... Args>
   void EmplaceItem(Args&&... args)
   {
+    if (!m_cancelled.IsSet())
     {
       std::lock_guard lg(m_lock);
       m_items.emplace(std::forward<Args>(args)...);
     }
     m_wakeup.Set();
   }
 
-private:
+  void Cancel() { m_cancelled.Set(); }
```

Comment:
I understand why you had to add `Cancel` functionality to stop additional work from piling up but I think the way you exposed it loosens the invariant of this little class too much. Now there is this potential "canceled" state where the thread is still running but no more work can be done.
The object is unusable after `Cancel` until you call `Reset` and I think it would simplify things if `Cancel` cleared and shutdown internally. I realize that makes `Cancel` a blocking call, but in its only use you call the blocking `Shutdown` anyways.
But if you really really want it this way I guess I don't care that much.

Reply:
Your suggestion is reasonable. I'll change it.

Reply:
Done.